### PR TITLE
[Graphbolt] Add MiniBatchBase

### LINKED ISSUE #6497: [Graphbolt] Feature fetcher support DGLMinibatch.
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #6497 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
@frozenbugs , @peizhou001 , I added a abstract class for `MiniBatch` and `DGLMiniBatch`. But I still need your help for several issues when fetching features.
1. Seems that `DGLBlock` does not have `original_edge_ids`. How can I retrieve the edge ids when fetching edge features for `DGLBlock`?
2. I moved as many attributes of `MiniBatch` and `DGLMiniBatch` to `MiniBatchBase` as I can. Please check whether there are missing attributes that are shared by the two classes.
3. To align with `DGLMiniBatch`, I renamed the `seed_nodes` in `MiniBatch` to `output_nodes`. I may need to refactor all the usages of `seed_nodes` in`MiniBatch` later. Please check whether this is reasonable.

File: python/dgl/graphbolt/minibatch.py [hunk 1]
```diff
@@ -62,12 +64,23 @@ class DGLMiniBatch:
...
     """
 
-    labels: Union[torch.Tensor, Dict[str, torch.Tensor]] = None
-    """Labels associated with seed nodes / node pairs in the graph.
-    - If `labels` is a tensor: It indicates the graph is homogeneous. The value
-      are corresponding labels to given 'output_nodes' or 'node_pairs'.
-    - If `labels` is a dictionary: The keys are node or edge type and the value
-      should be corresponding labels to given 'output_nodes' or 'node_pairs'.
+    @abstractmethod
+    def get_num_layers(self):
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/minibatch.py:68]: `num_layers()`?

**Comment by frozenbugs** [comment]
> @frozenbugs , @peizhou001 , I added a abstract class for `MiniBatch` and `DGLMiniBatch`. But I still need your help for several issues when fetching features.
> 
> 1. Seems that `DGLBlock` does not have `original_edge_ids`. How can I retrieve the edge ids when fetching edge features for `DGLBlock`?
> 2. I moved as many attributes of `MiniBatch` and `DGLMiniBatch` to `MiniBatchBase` as I can. Please check whether there are missing attributes that are shared by the two classes.
> 3. To align with `DGLMiniBatch`, I renamed the `seed_nodes` in `MiniBatch` to `output_nodes`. I may need to refactor all the usages of `seed_nodes` in`MiniBatch` later. Please check whether this is reasonable.

- The edge should be in dglblock.edata[dgl.EID], here is the [code](https://github.com/dmlc/dgl/blob/master/python/dgl/graphbolt/minibatch.py#L297C40-L297C40) that assigning the value. 
- Please don't rename the variable in MiniBatch to align with DGLMiniBatch, we intentionally kept them different, since the downstream of MiniBatch could be PyG or other library as well

File: python/dgl/graphbolt/minibatch.py (hunk 1 above)
Comment (peizhou001) [reply python/dgl/graphbolt/minibatch.py:68]: how about a property instead of method?

File: python/dgl/graphbolt/minibatch.py [hunk 2]
```diff
@@ -62,12 +64,23 @@ class DGLMiniBatch:
...
+        pass
+
+    @abstractmethod
+    def get_sampled_subgraphs(self):
```
Comment (peizhou001) [inline python/dgl/graphbolt/minibatch.py:72]: Ditto?

**Comment by czkkkkkk** [comment]
> Please don't rename the variable in MiniBatch to align with DGLMiniBatch, we intentionally kept them different, since the downstream of MiniBatch could be PyG or other library as well

I see. Do you think it is a better to keep `MiniBatchBase` as an interface class without any data, such as `input_nodes`, `node_features`, etc?

**Comment by czkkkkkk** [comment]
@frozenbugs, @Rhett-Ying, @peizhou001, I refactored the necessary interfaces of `MinibatchBase` to functions instead of properties. This is because defining them as properties needs to rename all the attributes of `MiniBatch` and `DGLMinibatch` to avoid naming conflict. For example, `input_nodes` will become `_input_nodes`. Then all the usages of these attributes, including constructors, will need to include the underscores. Take converting `MiniBatch` to `DGLMiniBatch` as an example.
```
minibatch = DGLMiniBatch(
    _blocks=self._to_dgl_blocks(),
    _node_features=self._node_features,
    _edge_features=self._edge_features,
    _labels=self._labels,
)
```
Please leave your comments if you have a better solution. If we think this solution is fine, I will go ahead to improve the docstring and add unittests.

File: python/dgl/graphbolt/minibatch.py [hunk 3]
```diff
@@ -15,6 +16,57 @@
...
+        """A representation of input nodes in the outermost layer. Conatins all
+        nodes in the 'sampled_subgraphs'.
+        - If `input_nodes` is a tensor: It indicates the graph is homogeneous.
+        - If `input_nodes` is a dictionary: The keys should be node type and the
+          value should be corresponding heterogeneous node id.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_num_layers(self) -> int:
```
Comment (frozenbugs) [inline python/dgl/graphbolt/minibatch.py:33]: ditto, omit get.

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**